service/history/replication/task_store.go

```diff
@@ -0,0 +1,223 @@
+// Once a cache becomes full it will not accept further messages for that cluster. Later those messages be fetched from DB and hydrated again.
+// A cache stores only a pointer to the message. It is hydrates once and shared across caches. Cluster acknowledging the message will remove it from that corresponding cache.
+// Once all clusters acknowledge it, no more references will be held, and GC will eventually pick it up.
+type TaskStore struct {
+	clusters      map[string]*Cache
+	domains       domainCache
+	hydrator      taskHydrator
+	rateLimiter   *quotas.DynamicRateLimiter
+	throttleRetry *backoff.ThrottleRetry
+
+	scope  metrics.Scope
+	logger log.Logger
+}
+
+type (
+	domainCache interface {
+		GetDomainByID(id string) (*cache.DomainCacheEntry, error)
+	}
+	taskHydrator interface {
+		Hydrate(ctx context.Context, task persistence.ReplicationTaskInfo) (*types.ReplicationTask, error)
+	}
+)
+
+// NewTaskStore create new instance of TaskStore
+func NewTaskStore(
+	config *config.Config,
+	clusterMetadata cluster.Metadata,
+	domains domainCache,
+	metricsClient metrics.Client,
+	logger log.Logger,
+	hydrator taskHydrator,
+) *TaskStore {
+
+	clusters := map[string]*Cache{}
+	for clusterName := range clusterMetadata.GetRemoteClusterInfo() {
+		clusters[clusterName] = NewCache(config.ReplicatorCacheCapacity)
+	}
+
+	retryPolicy := backoff.NewExponentialRetryPolicy(100 * time.Millisecond)
+	retryPolicy.SetMaximumAttempts(config.ReplicatorReadTaskMaxRetryCount())
+	retryPolicy.SetBackoffCoefficient(1)
+
+	return &TaskStore{
+		clusters: clusters,
+		domains:  domains,
+		hydrator: hydrator,
+		throttleRetry: backoff.NewThrottleRetry(
+			backoff.WithRetryPolicy(retryPolicy),
+			backoff.WithRetryableError(persistence.IsTransientError),
+		),
+		scope:       metricsClient.Scope(metrics.ReplicatorCacheManagerScope),
+		logger:      logger.WithTags(tag.ComponentReplicationCacheManager),
+		rateLimiter: quotas.NewDynamicRateLimiter(config.ReplicationTaskGenerationQPS.AsFloat64()),
+	}
+}
+
+// Get will return a hydrated replication message for a given cluster based on raw task info.
+// It will either return it immediately from cache or hydrate it, store in cache and then return.
+//
+// Returned task may be nil. This may be due domain not existing in a given cluster or replication message is not longer relevant.
+// Either case is valid and such replication message should be ignored and not returned in the response.
+func (m *TaskStore) Get(ctx context.Context, cluster string, info persistence.ReplicationTaskInfo) (*types.ReplicationTask, error) {
+	cache, ok := m.clusters[cluster]
+	if !ok {
+		return nil, errUnknownCluster
+	}
+
+	domain, err := m.domains.GetDomainByID(info.DomainID)
+	if err != nil {
+		return nil, fmt.Errorf("resolving domain: %w", err)
+	}
+
+	// Domain does not exist in this cluster, do not replicate the task
+	if !domain.HasReplicationCluster(cluster) {
+		return nil, nil
```

Comment:
Is this an expected scenario ever?

Reply:
It is. Domain can be configured on only on a subset of clusters. No need to replicate them everywhere. Although this is probably not very common.